extensions/amp-lightbox-gallery/0.1/amp-lightbox-gallery.js

```diff
@@ -694,59 +689,86 @@ export class AmpLightboxGallery extends AMP.BaseElement {
   /**
    * This function verifies that the source element is an amp-img and contains
    * an img element and preserves the natural aspect ratio of the original img.
-   * @param {!Element} element
-   * @return {boolean}
+   * @param {!Element|null} element
+   * @return {!Promise<boolean>}
    * @private
    */
   shouldAnimate_(element) {
+    if (!element || !isLoaded(element)) {
+      return Promise.resolve(false);
+    }
     if (!ELIGIBLE_TAP_TAGS[element.tagName]) {
-      return false;
+      return Promise.resolve(false);
     }
     const img = elementByTag(dev().assertElement(element), 'img');
     if (!img) {
-      return false;
+      return Promise.resolve(false);
     }
-    const naturalAspectRatio = img.naturalWidth / img.naturalHeight;
-    const elementHeight = element./*OK*/offsetHeight;
-    const elementWidth = element./*OK*/offsetWidth;
-    const ampImageAspectRatio = elementWidth / elementHeight;
-    return Math.abs(naturalAspectRatio - ampImageAspectRatio) < EPSILON;
+    return this.vsync_.measurePromise(() => {
+      const naturalAspectRatio = img.naturalWidth / img.naturalHeight;
+      const elementHeight = element./*OK*/offsetHeight;
+      const elementWidth = element./*OK*/offsetWidth;
+      const ampImageAspectRatio = elementWidth / elementHeight;
+      return Math.abs(naturalAspectRatio - ampImageAspectRatio) < EPSILON;
+    });
   }
 
   /**
-   * Entry animation to transition in a lightboxable image
-   * @return {!Promise}
+   * @return {!Promise<boolean>}
    * @private
    */
-  // TODO (cathyxz): make this generalizable to more than just images
-  enter_() {
+  shouldExit_() {
+    const element = this.getCurrentElement_().sourceElement;
+    return this.shouldAnimate_(element)
+        .then(shouldAnimate => {
+          return shouldAnimate && (element == this.sourceElement_
+          || this.manager_.hasCarousel(this.currentLightboxGroupId_));
```

Comment:
nit: Maybe use a named variable to describe the semantics of `element == this.sourceElement_ || this.manager_.hasCarousel(this.currentLightboxGroupId_)`. It's kind of hard for me to grok why that condition means the lightbox gallery should exit.

Reply:
Yup.